single_kernel_mongo/core/operator.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Abstract Operator for Mongo Related Charms."""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, ClassVar
+
+from ops.charm import RelationDepartedEvent
+from ops.framework import Object
+from ops.model import Unit
+
+from single_kernel_mongo.config.literals import CharmRole, Substrates
+from single_kernel_mongo.managers.config import CommonConfigManager
+from single_kernel_mongo.managers.mongo import MongoManager
+from single_kernel_mongo.state.charm_state import CharmState
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.abstract_charm import AbstractMongoCharm
+    from single_kernel_mongo.core.workload import WorkloadBase
+    from single_kernel_mongo.events.database import DatabaseEventsHandler
+    from single_kernel_mongo.events.tls import TLSEventsHandler
+    from single_kernel_mongo.managers.tls import TLSManager
+
+
+class OperatorProtocol(ABC, Object):
```

Comment:
This is not about the `AbstractMongoCharm` but about the operators (`MongoDBOperator` and `MongosOperator`). This is likely to be updated if there is some shared code between the two operators.
This class is here only for typing purposes and enforcing an interface for the two operators (and more if we need to bring some new operators to Mongo Charms.

The code structure is:
```
charm -- > MongoDBOperator
      |-> MongosOperator
```

If there is some shared code to land on events it will be either in the OperatorProtocol or in the `AbstractCharm` (so preceding the code run in the events of the Operators).